### PR TITLE
References to mutable values

### PR DESCRIPTION
**Pull request by ironcev**
## Description

This PR implements references to mutable values, `&mut T`, as defined in [references](https://github.com/FuelLabs/sway-rfcs/blob/ironcev/amend-references/files/0010-references.sw). The overall effort related to references is tracked in #5063.

References to mutable values:
- can be created using `&mut`
- can be used in type system wherever a type is expected
- coerce to references (`&mut T` -> `&T`) and thus can be used wherever "regular" references are expected
- can be passed to and returned from functions

The last point is a step in direction of replacing `ref mut` function parameters with either `&` or `&mut`.

References to mutable values cannot be taken on constants and immutable variables (see errors below). (Note that we slightly departure from the Rust counterparts here. Rust allows `&mut` on constants but issues a warning.)

Additionally, the PR implements `Diagnostic` for the `ConflictingImplsForTraitAndType` and shows the already existing, conflicting implementation of the trait.

## Demo

For more demos, see the tests in this PR.

```Sway
let mut x = 123u32;
let r_m_x: &mut u32 = &mut x;

fn fun(p: &mut T) -> &mut T { ... }

impl Foo for &mut &mut T { ... }

type RefMutRefMut = &mut &mut u64;
```
References to mutable values cannot reference constants

Trait is already implemented for type

## Checklist

- [x] I have linked to any relevant issues.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [ ] I have updated the documentation where relevant (API docs, the reference, and the Sway book).
- [x] I have added tests that prove my fix is effective or that my feature works.
- [ ] I have added (or requested a maintainer to add) the necessary `Breaking*` or `New Feature` labels where relevant.
- [x] I have done my best to ensure that my PR adheres to [the Fuel Labs Code Review Standards](https://github.com/FuelLabs/rfcs/blob/master/text/code-standards/external-contributors.md).
- [x] I have requested a review from the relevant team or maintainers.

### CODE REVIEW

File: test/src/e2e_vm_tests/test_programs/should_pass/language/references/dereferencing_operator_dot_on_structs/test.toml [hunk 1]
```diff
@@ -1,4 +1,4 @@
 category = "run"
 expected_result = { action = "return", value = 42 }
-validate_abi = true
+validate_abi = false
```
Comment (sdankel) [inline test/src/e2e_vm_tests/test_programs/should_pass/language/references/dereferencing_operator_dot_on_structs/test.toml:3]: Why did this change? I see it's changed for many of the test programs.

File: sway-core/src/semantic_analysis/ast_node/expression/typed_expression.rs [hunk 2]
```diff
@@ -2136,9 +2170,18 @@ impl ty::TyExpression {
         // reference to the expected type.
         // Otherwise, we pass a new `TypeInfo::Unknown` as the annotation, to allow the `expr`
         // to be evaluated without any expectations.
+        // Since `&mut T` coerces into `&T` we always go with a lesser expectation, `&T`.
+        // Thus, `to_mutable_vale` is set to false.
```
Comment (sdankel) [inline sway-core/src/semantic_analysis/ast_node/expression/typed_expression.rs:2174]:
```suggestion
        // Thus, `to_mutable_value` is set to false.
```

**Review by sdankel — COMMENTED** [review]
Awesome work!

File: sway-core/src/semantic_analysis/ast_node/expression/typed_expression.rs (hunk 2 above)
Comment (ironcev) [reply sway-core/src/semantic_analysis/ast_node/expression/typed_expression.rs]: Done.

File: test/src/e2e_vm_tests/test_programs/should_pass/language/references/dereferencing_operator_dot_on_structs/test.toml (hunk 1 above)
Comment (ironcev) [reply test/src/e2e_vm_tests/test_programs/should_pass/language/references/dereferencing_operator_dot_on_structs/test.toml:3]: The tests are actually not checking/stressing/testing ABI at all. Like expected functions, types, configurables, etc. There is nothing special in the ABIs, just `main` function and used types.

The question is actually why the `validate_abi` was set to `true` at all :wink: Well, as usually the fact, I got the first test back then by copy-pasting some existing one that was by chance validating ABI and I didn't remove the validation...

File: sway-core/src/semantic_analysis/namespace/trait_map.rs [hunk 3]
```diff
@@ -209,7 +209,60 @@ impl TraitMap {
...
-                let types_are_subset = unify_checker.check(type_id, *map_type_id);
+
+                // Types are subset if the `type_id` that we want to insert can unify with the
+                // existing `map_type_id`. In addition we need to additionally check for the case of
+                // `&mut <type>` and `&<type>`.
+                let types_are_subset = if unify_checker.check(type_id, *map_type_id) {
+                    is_unified_type_subset(engines.te(), type_id, *map_type_id)
+                } else {
+                    false
+                };
```
Comment (IGI-111) [inline sway-core/src/semantic_analysis/namespace/trait_map.rs:220]:
```suggestion
                let types_are_subset = unify_checker.check(type_id, *map_type_id) && is_unified_type_subset(engines.te(), type_id, *map_type_id);
```

File: sway-error/src/error.rs [hunk 4]
```diff
@@ -1861,6 +1879,63 @@ impl ToDiagnostic for CompileError {
...
+                    Hint::info(
+                        source_engine,
+                        existing_impl_span.clone(),
+                        format!("This is the already existing implementation of \"{}\" for \"{type_implementing_for}\".",
+                            call_path_suffix_with_args(trait_name)
+                        )
+                    ),
+                ],
+                help: vec![
+                    format!("In Sway, there can be at most one implementation of a trait for any given type."),
```
Comment (IGI-111) [inline sway-error/src/error.rs:1935]:
```suggestion
                    "In Sway, there can be at most one implementation of a trait for any given type.".to_string(),
```

File: sway-error/src/error.rs [hunk 5]
```diff
@@ -1861,6 +1879,63 @@ impl ToDiagnostic for CompileError {
...
+                    format!("This property is called \"trait coherence\"."),
```
Comment (IGI-111) [inline sway-error/src/error.rs:1936]:
```suggestion
                    "This property is called \"trait coherence\".".to_string(),
```

**Review by IGI-111 — APPROVED** [review]

**Review by xunilrj — APPROVED** [review]

**Pull request merged**